packages/langium/src/grammar/type-system/inferred-types.ts

```diff
@@ -410,14 +410,13 @@ function calculateAst(alternatives: TypeAlternative[]): InterfaceType[] {
 }
 
 function flattenTypes(alternatives: TypeAlternative[]): TypeAlternative[] {
-    const names = new Set<string>(alternatives.map(e => e.name));
+    const nameToAlternatives = alternatives.reduce((acc, e) => { acc.add(e.name, e); return acc; }, new MultiMap<string, TypeAlternative>());
```

Comment:
It looks like you can just return `acc.add(e.name, e)` directly, as it seems to return instance of itself. Could probably shorten it to 
`(acc, e) => acc.add(e.name, e), ...`.